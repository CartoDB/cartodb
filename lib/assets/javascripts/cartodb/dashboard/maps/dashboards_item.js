var cdb = require('cartodb.js');
var moment = require('moment');
var navigateThroughRouter = require('../../common/view_helpers/navigate_through_router');
var Utils = require('cdb.Utils');
cdb.admin = require('cdb.admin');

/**
 * Represents a map card on dashboard.
 */
module.exports = cdb.core.View.extend({

  className: 'MapsList-item',
  tagName: 'li',

  initialize: function() {
    this.router = this.options.router;
    this.user = this.options.user;
    this.template = cdb.templates.getTemplate('dashboard/views/dashboards_item');
  },

  render: function() {
    this.clearSubViews();
    var url = this.model.dashboardURL();

    this.$el.html(
      this.template({
        url: url,
        name: this.model.get('name'),
        description: this.model.get('description'),
        timeDiff: moment(this.model.get('updated_at')).fromNow(),
      })
    );

    return this;
  }

});
