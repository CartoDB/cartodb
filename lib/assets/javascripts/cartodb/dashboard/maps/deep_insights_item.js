var cdb = require('cartodb.js-v3');
var moment = require('moment-v3');
var navigateThroughRouter = require('../../common/view_helpers/navigate_through_router');
var Utils = require('cdb.Utils');
cdb.admin = require('cdb.admin');

/**
 * Represents a map card on dashboard.
 */
module.exports = cdb.core.View.extend({

  className: 'MapsList-item',
  tagName: 'li',

  initialize: function() {
    this.router = this.options.router;
    this.user = this.options.user;
    this.template = cdb.templates.getTemplate('dashboard/views/deep_insights_item');
  },

  render: function() {
    this.clearSubViews();
    var url = this.model.deepInsightsUrl(this.user);

    this.$el.html(
      this.template({
        url: url,
        title: this.model.get('title'),
        description: this.model.get('description'),
        timeDiff: moment(this.model.get('updated_at')).fromNow(),
      })
    );

    return this;
  }

});
