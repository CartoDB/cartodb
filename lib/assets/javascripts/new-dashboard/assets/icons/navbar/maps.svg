<svg xmlns="http://www.w3.org/2000/svg" width="19" height="21">
  <path d="M18.555 15.832l-6.006 3.992c-.154.103-.326.176-.549.176-.223 0-.444-.074-.624-.219l-4.429-3.543-5.392 3.594C1.387 19.944 1.162 20 1 20c-.162 0-.324-.038-.472-.118C.203 19.708 0 19.369 0 19V5c0-.333.167-.646.445-.832l6-4C6.6.066 6.82 0 7.035.008c.215.008.44.094.59.211l4.429 3.544L17.446.168c.306-.205.701-.222 1.026-.049.325.173.528.512.528.881v14c0 .334-.167.647-.445.832zM2 5.535v11.597l4-2.667V2.869zm9 11.385V5.481L8 3.082V14.52zm6-2.455V2.869l-4 2.666v11.597z" fill="#fff"/>
  <path class="navbar-iconFill" d="M12 17.5V4.274L6 1.5v13.225z" fill="transparent"/>
</svg>