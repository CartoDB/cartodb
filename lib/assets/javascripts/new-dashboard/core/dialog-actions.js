// Dialogs
import Dialog from 'new-dashboard/components/Backbone/Dialog.vue';
import CreateDialog from 'new-dashboard/components/Backbone/Dialogs/CreateDialog';
import ChangePrivacy from 'new-dashboard/components/Backbone/Dialogs/ChangePrivacy';
import DeleteDialog from 'new-dashboard/components/Backbone/Dialogs/DeleteDialog';
import DuplicateMap from 'new-dashboard/components/Backbone/Dialogs/DuplicateMap';
import LockVisualization from 'new-dashboard/components/Backbone/Dialogs/LockVisualization';
import MapMetadata from 'new-dashboard/components/Backbone/Dialogs/MapMetadata';
import DatasetMetadata from 'new-dashboard/components/Backbone/Dialogs/DatasetMetadata';
import Share from 'new-dashboard/components/Backbone/Dialogs/Share';

export function editMapMetadata (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <MapMetadata
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, MapMetadata },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function editDatasetMetadata (dataset, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <DatasetMetadata
          :dataset="dataset"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['dataset'],
      components: { Dialog, DatasetMetadata },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { dataset }
  );
}

export function createMap (selectedDatasets, backgroundPollingView, mamufasImportView) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <CreateDialog :selectedItems="selectedItems" :backgroundPollingView="backgroundPollingView" :mamufasImportView="mamufasImportView" @close="$emit('close')"/>
      </Dialog>`,
      props: ['selectedItems', 'backgroundPollingView', 'mamufasImportView'],
      components: { Dialog, CreateDialog }
    },
    {
      selectedItems: selectedDatasets,
      backgroundPollingView,
      mamufasImportView
    }
  );
}

export function changePrivacy (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <ChangePrivacy
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, ChangePrivacy },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function shareVisualization (visualization, actionHandlers) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <Share
          :visualization="visualization"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, Share },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    },
    { visualization }
  );
}

export function duplicateVisualization (visualization) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DuplicateMap :visualization="visualization" @close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, DuplicateMap }
    },
    { visualization }
  );
}

export function changeLockState (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockVisualization
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, LockVisualization },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function changeVisualizationsLockState (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockVisualization
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, LockVisualization },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function deleteVisualization (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}

export function deleteVisualizations (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          v-on="getEventListeners()" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        getEventListeners: function () {
          return actionHandlers || {};
        }
      }
    }, dialogProps
  );
}
