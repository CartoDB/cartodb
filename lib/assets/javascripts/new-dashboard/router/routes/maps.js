// Hooks
import { mapsBeforeEnter, externalMapsBeforeEnter } from 'new-dashboard/router/hooks/check-navigation';

// Lazy Pages
const Maps = () => import('new-dashboard/pages/Maps/Maps');
const Carto = () => import('new-dashboard/pages/Maps/Carto');
const External = () => import('new-dashboard/pages/Maps/Carto');

const routes = [
  {
    path: '/maps/:filter?',
    component: Maps,
    meta: {
      title: () => 'Maps | CARTO'
    },
    children: [
      {
        path: '',
        name: 'maps',
        component: Carto,
        meta: {
          title: () => 'Maps | CARTO'
        },
        beforeEnter: mapsBeforeEnter
      },
      {
        path: 'external',
        name: 'external',
        component: External,
        meta: {
          title: () => 'Kepler.gl Maps | CARTO'
        },
        beforeEnter: externalMapsBeforeEnter
      }
    ]
  },
  {
    path: '/maps/external/:filter',
    name: 'external_filtered',
    component: Maps,
    meta: {
      title: () => 'Kepler.gl  Maps | CARTO'
    },
    children: [
      {
        path: '',
        name: 'external_filtered',
        component: External,
        meta: {
          title: () => 'Kepler.gl  Maps | CARTO'
        },
        beforeEnter: externalMapsBeforeEnter
      }
    ]
  }
];

export default routes;
