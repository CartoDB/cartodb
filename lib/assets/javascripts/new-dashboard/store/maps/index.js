import CartoNode from 'carto-node';
import toObject from 'new-dashboard/utils/to-object';
import featuredFavoritedMaps from 'new-dashboard/store/maps/featured-favorited-maps';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/filters';

const client = new CartoNode.AuthenticatedClient();

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const maps = {
  namespaced: true,
  modules: {
    featuredFavoritedMaps
  },
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 12
  },
  getters: {},
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setMaps (state, maps) {
      state.list = toObject(maps.visualizations, 'id');
      state.metadata = {
        total_entries: maps.total_entries,
        total_likes: maps.total_likes,
        total_shared: maps.total_shared,
        total_locked: maps.total_locked,
        total_user_entries: maps.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setMapAttributes (state, {mapId, mapAttributes}) {
      Object.assign(state.list[mapId], mapAttributes);
    },

    setOrder (state, orderOpts) {
      state.order = orderOpts.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOpts.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage = 12) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike (state, {mapId, liked}) {
      state.list[mapId].liked = liked;
    },

    updateNumberLikes (state, {mapId, likes}) {
      state.list[mapId].likes = likes;
    }
  },
  actions: {
    fetchMaps (context) {
      context.commit('setFetchingState');
      const params = {
        ...Filters[context.state.filterType],
        types: 'derived',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage
      };

      client.getVisualization('',
        params,
        function (err, _, data) {
          if (err) {
            context.commit('setRequestError', err);
            return;
          }
          context.commit('setMaps', data);
          context.commit('setPagination', context.state.page);
        }
      );
    },
    goToPage (context, page) {
      context.commit('setPagination', page);
      context.dispatch('fetchMaps');
    },
    filterMaps (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetchMaps');
    },
    orderMaps (context, orderOpts) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOpts);
      context.dispatch('fetchMaps');
    },
    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', {order: DEFAULT_VALUES.order, orderDirection: DEFAULT_VALUES.orderDirection});
    },
    setPerPage (context, resultsPerPage) {
      context.commit('setResultsPerPage', resultsPerPage);
    },
    updateMap (context, mapOptions) {
      context.commit('setMapAttributes', mapOptions);
    },
    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', {direction: options.order_direction, order: options.order});
      context.dispatch('fetchMaps');
    },
    like (context, map) {
      const currentLikeStatus = map.liked;
      context.commit('updateLike', { mapId: map.id, liked: true });
      client.like(map.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { mapId: map.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { mapId: map.id, likes: data.likes });
        }
      );
    },
    deleteLike (context, map) {
      const currentLikeStatus = map.liked;
      context.commit('updateLike', { mapId: map.id, liked: false });
      client.deleteLike(map.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { mapId: map.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { mapId: map.id, likes: data.likes });
        }
      );
    }
  }
};

export default maps;
