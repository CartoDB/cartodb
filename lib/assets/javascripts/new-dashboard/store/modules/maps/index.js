import toObject from 'new-dashboard/utils/to-object';
import featuredFavoritedMaps from 'new-dashboard/store/modules/maps/featured-favorited-maps';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/configuration/filters';

import * as VisualizationActions from '../../actions/visualizations';
import * as VisualizationMutations from '../../mutations/visualizations';

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const maps = {
  namespaced: true,
  modules: {
    featuredFavoritedMaps
  },
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 12
  },
  getters: {},
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setVisualizations (state, maps) {
      state.list = toObject(maps.visualizations, 'id');
      state.metadata = {
        total_entries: maps.total_entries,
        total_likes: maps.total_likes,
        total_shared: maps.total_shared,
        total_locked: maps.total_locked,
        total_user_entries: maps.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setOrder (state, orderOpts) {
      state.order = orderOpts.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOpts.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage = 12) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike: VisualizationMutations.updateLike,
    updateNumberLikes: VisualizationMutations.updateNumberLikes
  },
  actions: {
    fetch (context) {
      const params = {
        ...Filters[context.state.filterType],
        types: 'derived',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage
      };

      VisualizationActions.fetchVisualizations(context, params);
    },
    goToPage (context, page) {
      context.commit('setPagination', page);
      context.dispatch('fetch');
    },
    filterMaps (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetch');
    },
    orderMaps (context, orderOpts) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOpts);
      context.dispatch('fetch');
    },
    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', {order: DEFAULT_VALUES.order, orderDirection: DEFAULT_VALUES.orderDirection});
    },
    setPerPage (context, resultsPerPage) {
      context.commit('setResultsPerPage', resultsPerPage);
    },
    updateVisualization: VisualizationActions.updateVisualization,
    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', {direction: options.order_direction, order: options.order});
      context.dispatch('fetch');
    },
    like: VisualizationActions.like,
    deleteLike: VisualizationActions.deleteLike
  }
};

export default maps;
