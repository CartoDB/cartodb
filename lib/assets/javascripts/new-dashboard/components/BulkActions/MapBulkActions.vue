<template>
  <BulkActions :actions="actions[actionMode]" v-on="getEventListeners()"></BulkActions>
</template>

<script>
import BulkActions from 'new-dashboard/components/BulkActions/BulkActions';
import * as DialogActions from 'new-dashboard/core/dialog-actions';

export default {
  name: 'MapBulkActions',
  components: {
    BulkActions
  },
  props: {
    areAllMapsSelected: {
      type: Boolean,
      default: false
    },
    selectedMaps: {
      type: Array,
      required: true
    }
  },
  computed: {
    actions () {
      return {
        single: [
          { name: this.$t('BulkActions.maps.selectAllMaps'), event: 'selectAll' },
          { name: this.$t('BulkActions.maps.changeMapPrivacy'), event: 'changePrivacy' },
          { name: this.$t('BulkActions.maps.duplicateMap'), event: 'duplicateMap' },
          { name: this.$t('BulkActions.maps.lockMap'), event: 'lockMap' },
          { name: this.$t('BulkActions.maps.deleteMap'), event: 'deleteMap', isDestructive: true }
        ],
        multiple: [
          { name: this.$t('BulkActions.maps.selectAllMaps'), event: 'selectAll', shouldBeHidden: this.areAllMapsSelected },
          { name: this.$t('BulkActions.maps.deselectAllMaps'), event: 'deselectAll' },
          { name: this.$t('BulkActions.maps.lockMaps'), event: 'lockMaps' },
          { name: this.$t('BulkActions.maps.deleteMaps'), event: 'deleteMaps', isDestructive: true }
        ],
        lock: [
          { name: this.$t('BulkActions.maps.selectAllMaps'), event: 'selectAll', shouldBeHidden: this.areAllMapsSelected },
          { name: this.$t('BulkActions.maps.unlockMap'), event: 'unlockMap' }
        ],
        multipleLock: [
          { name: this.$t('BulkActions.maps.deselectAllMaps'), event: 'deselectAll' },
          { name: this.$t('BulkActions.maps.unlockMaps'), event: 'unlockMaps' }
        ]
      };
    },
    actionMode () {
      const isAnyMapLocked = this.selectedMaps.filter(map => map.locked);

      if (isAnyMapLocked.length) {
        return isAnyMapLocked.length > 1 ? 'multipleLock' : 'lock';
      }

      return this.selectedMaps.length > 1 ? 'multiple' : 'single';
    }
  },
  methods: {
    getActionHandlers () {
      return {
        deselectAll: () => {
          this.deselectAll();
        },
        fetchList: () => {
          this.$store.dispatch('maps/fetchMaps');
        },
        updateVisualization: (model) => {
          this.$store.dispatch('maps/updateMap', { mapId: model.get('id'), mapAttributes: model.attributes });
        }
      };
    },
    getEventListeners () {
      const events = this.actions[this.actionMode].map(action => action.event);

      return events.reduce(
        (eventListeners, action) => {
          eventListeners[action] = this[action].bind(this);
          return eventListeners;
        }, {}
      );
    },
    showModal (componentDefinition, componentPropsData) {
      this.$modal.show(
        componentDefinition,
        componentPropsData,
        { width: '100%', height: '100%' }
      );
    },
    selectAll () {
      this.$emit('selectAll');
    },
    deselectAll () {
      this.$emit('deselectAll');
    },
    changePrivacy () {
      DialogActions.changePrivacy.apply(this, [this.selectedMaps[0]]);
    },
    duplicateMap () {
      DialogActions.duplicateVisualization.apply(this, [this.selectedMaps[0]]);
    },
    unlockMap () {
      DialogActions.changeLockState.apply(this, [
        this.selectedMaps[0],
        'maps',
        this.getActionHandlers()
      ]);
    },
    lockMap () {
      DialogActions.changeLockState.apply(this, [
        this.selectedMaps[0],
        'maps',
        this.getActionHandlers()
      ]);
    },
    deleteMap () {
      DialogActions.deleteVisualization.apply(this, [
        this.selectedMaps[0],
        'maps',
        this.getActionHandlers()
      ]);
    },
    unlockMaps () {
      DialogActions.changeVisualizationsLockState.apply(this, [
        this.selectedMaps,
        'maps',
        this.getActionHandlers()
      ]);
    },
    lockMaps () {
      DialogActions.changeVisualizationsLockState.apply(this, [
        this.selectedMaps,
        'maps',
        this.getActionHandlers()
      ]);
    },
    deleteMaps () {
      DialogActions.deleteVisualizations.apply(this, [
        this.selectedMaps,
        'maps',
        this.getActionHandlers()
      ]);
    }
  }
};
</script>
