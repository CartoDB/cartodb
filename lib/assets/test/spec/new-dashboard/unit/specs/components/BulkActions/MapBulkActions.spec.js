import { mount } from '@vue/test-utils';
import MapBulkActions from 'new-dashboard/components/BulkActions/MapBulkActions';
import maps from '../../fixtures/visualizations';
import mapsLocked from '../../fixtures/visualizationsLocked';
import * as DialogActions from '../__mocks__/dialog-actions';

const $t = key => key;

describe('MapBulkActions.vue', () => {
  it('should render all bulk actions', () => {
    const mapBulkAction = mount(MapBulkActions, {
      propsData: {
        areAllMapsSelected: false,
        selectedMaps: []
      },
      mocks: {
        $t
      }
    });
    expect(mapBulkAction).toMatchSnapshot();
  });

  it('should render all multiple actions for all selected "Deselect all", "Lock maps" and "Delete maps" actions', () => {
    const data = maps.visualizations;
    const mapBulkAction = mount(MapBulkActions, {
      propsData: {
        areAllMapsSelected: true,
        selectedMaps: data
      },
      mocks: {
        $t
      }
    });
    expect(mapBulkAction).toMatchSnapshot();
  });

  it('should render all single actions: "Select all", "Create map", "Change privacy", "Duplicate map", "Lock map", "Delete map"', () => {
    const data = maps.visualizations.slice(0, 1);
    const mapBulkAction = mount(MapBulkActions, {
      propsData: {
        areAllMapsSelected: false,
        selectedMaps: data
      },
      mocks: {
        $t
      }
    });
    expect(mapBulkAction).toMatchSnapshot();
  });

  it('should render all multiple actions: "Select all", "Deselect all", "Lock maps", "Delete maps"', () => {
    const data = maps.visualizations;
    const mapBulkAction = mount(MapBulkActions, {
      propsData: {
        areAllMapsSelected: false,
        selectedMaps: data
      },
      mocks: {
        $t
      }
    });
    expect(mapBulkAction).toMatchSnapshot();
  });

  it('should render single locked action: "Unlock map"', () => {
    const data = mapsLocked.visualizations;
    const mapBulkAction = mount(MapBulkActions, {
      propsData: {
        areAllMapsSelected: false,
        selectedMaps: data
      },
      mocks: {
        $t
      }
    });
    expect(mapBulkAction).toMatchSnapshot();
  });

  it('should render multiple locked action: "Unlock maps"', () => {
    const data = mapsLocked.visualizations.slice(0, 2);
    const mapBulkAction = mount(MapBulkActions, {
      propsData: {
        areAllMapsSelected: false,
        selectedMaps: data
      },
      mocks: {
        $t
      }
    });
    expect(mapBulkAction).toMatchSnapshot();
  });

  describe('computed', () => {
    it('should return "multiple" action name when more than one map is selected', () => {
      const data = maps.visualizations;
      const mapBulkAction = mount(MapBulkActions, {
        propsData: {
          selectedMaps: data
        },
        mocks: {
          $t
        }
      });

      const result = mapBulkAction.vm.actionMode;
      expect(result).toBe('multiple');
    });

    it('should return "single" action name when there is only a map selected', () => {
      const data = maps.visualizations.slice(0, 1);
      const mapBulkAction = mount(MapBulkActions, {
        propsData: {
          selectedMaps: data
        },
        mocks: {
          $t
        }
      });
      const result = mapBulkAction.vm.actionMode;
      expect(result).toBe('single');
    });

    it('should return "lock" action name when there is only a locked map selected', () => {
      const data = mapsLocked.visualizations.slice(0, 1);
      const mapBulkAction = mount(MapBulkActions, {
        propsData: {
          selectedMaps: data
        },
        mocks: {
          $t
        }
      });
      const result = mapBulkAction.vm.actionMode;
      expect(result).toBe('lock');
    });

    it('should return "multipleLock" action name when there are more than one locked map selected', () => {
      const data = mapsLocked.visualizations;
      const mapBulkAction = mount(MapBulkActions, {
        propsData: {
          selectedMaps: data
        },
        mocks: {
          $t
        }
      });
      const result = mapBulkAction.vm.actionMode;
      expect(result).toBe('multipleLock');
    });
  });

  describe('methods', () => {
    let $modal;
    let data;
    let mapBulkAction;
    beforeEach(function () {
      $modal = {
        show: jest.fn()
      };
      data = maps.visualizations;
      mapBulkAction = mount(MapBulkActions, {
        propsData: {
          selectedMaps: data
        },
        mocks: {
          $modal,
          $t,
          backboneViews: {
            backgroundPollingView: {
              getBackgroundPollingView () {
                return {};
              }
            }
          }
        }
      });
    });

    it('should return all events listened', () => {
      const events = {
        'deleteMaps': {},
        'deselectAll': {},
        'lockMaps': {},
        'selectAll': {}
      };
      expect(mapBulkAction.vm.getEventListeners()).toMatchObject(events);
    });

    it('should call the function to show modal', () => {
      mapBulkAction.vm.showModal();
      expect($modal.show).toHaveBeenCalled();
    });

    it('should call the function selectAll', () => {
      mapBulkAction.vm.selectAll();
      expect(mapBulkAction.emitted('selectAll').length).toBe(1);
    });

    it('should call the function deselectAll', () => {
      mapBulkAction.vm.deselectAll();
      expect(mapBulkAction.emitted('deselectAll').length).toBe(1);
    });

    it('should open change privacy modal', () => {
      mapBulkAction.vm.changePrivacy();
      expect(DialogActions.changePrivacy).toHaveBeenCalled();
    });

    it('should open duplicate map modal', () => {
      mapBulkAction.vm.duplicateMap();
      expect(DialogActions.duplicateVisualization).toHaveBeenCalled();
    });

    it('should open change visualizations lock state dialog via unlockMaps funcion', () => {
      mapBulkAction.vm.unlockMaps();
      expect(DialogActions.changeVisualizationsLockState).toHaveBeenCalled();
    });

    it('should open change lock state dialog via lockMap function', () => {
      mapBulkAction.vm.lockMap();
      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should open change visualizations lock state dialog via lockMaps funcion', () => {
      mapBulkAction.vm.lockMaps();
      expect(DialogActions.changeVisualizationsLockState).toHaveBeenCalled();
    });

    it('should open change lock state dialog via unlockMap function', () => {
      mapBulkAction.vm.unlockMap();
      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should open change lock state dialog via unlockMaps function', () => {
      mapBulkAction.vm.unlockMaps();
      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should open delete visualization dialog', () => {
      const spy = spyOn(mapBulkAction.vm, 'deselectAll');
      mapBulkAction.vm.deleteMap();
      expect(DialogActions.deleteVisualization).toHaveBeenCalled();
      expect(spy).toHaveBeenCalled();
    });

    it('should open delete visualizations dialog', () => {
      mapBulkAction.vm.deleteMaps();
      expect(DialogActions.deleteVisualizations).toHaveBeenCalled();
    });
  });
});
