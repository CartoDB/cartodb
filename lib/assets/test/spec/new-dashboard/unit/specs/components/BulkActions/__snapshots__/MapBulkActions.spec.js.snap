// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapBulkActions.vue should render all bulk actions 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.selectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.changeMapPrivacy
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.duplicateMap
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.lockMap
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.maps.deleteMap
  </button></section>
`;

exports[`MapBulkActions.vue should render all multiple actions for all selected "Deselect all", "Lock maps" and "Delete maps" actions 1`] = `
<section class="bulk-actions">
  <!----><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.deselectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.lockMaps
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.maps.deleteMaps
  </button></section>
`;

exports[`MapBulkActions.vue should render all multiple actions: "Select all", "Deselect all", "Lock maps", "Delete maps" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.selectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.deselectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.lockMaps
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.maps.deleteMaps
  </button></section>
`;

exports[`MapBulkActions.vue should render all single actions: "Select all", "Create map", "Change privacy", "Duplicate map", "Lock map", "Delete map" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.selectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.changeMapPrivacy
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.duplicateMap
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.lockMap
  </button><button class="bulk-actions__button button button--ghost is-txtAlert">
    BulkActions.maps.deleteMap
  </button></section>
`;

exports[`MapBulkActions.vue should render multiple locked action: "Unlock maps" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.deselectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.unlockMaps
  </button></section>
`;

exports[`MapBulkActions.vue should render single locked action: "Unlock map" 1`] = `
<section class="bulk-actions"><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.deselectAllMaps
  </button><button class="bulk-actions__button button button--ghost">
    BulkActions.maps.unlockMaps
  </button></section>
`;
